File: clients/rust/crates/rubin-consensus/src/featurebits.rs
use crate::constants::{SIGNAL_THRESHOLD, SIGNAL_WINDOW};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum FeatureBitState {
    Defined,
    Started,
    LockedIn,
    Active,
    Failed,
}

impl FeatureBitState {
    pub fn as_str(&self) -> &'static str {
        match self {
            FeatureBitState::Defined => "DEFINED",
            FeatureBitState::Started => "STARTED",
            FeatureBitState::LockedIn => "LOCKED_IN",
            FeatureBitState::Active => "ACTIVE",
            FeatureBitState::Failed => "FAILED",
        }
    }
}

#[derive(Clone, Debug)]
pub struct FeatureBitDeployment {
    pub name: String,
    pub bit: u8,
    pub start_height: u64,
    pub timeout_height: u64,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct FeatureBitEval {
    pub state: FeatureBitState,
    pub boundary_height: u64,
    pub prev_window_signal_count: u32,
    pub signal_window: u64,
    pub signal_threshold: u32,
}

fn next_state(
    prev: FeatureBitState,
    boundary_height: u64,
    prev_window_signal_count: u32,
    d: &FeatureBitDeployment,
) -> FeatureBitState {
    match prev {
        FeatureBitState::Defined => {
            if boundary_height >= d.start_height {
                FeatureBitState::Started
            } else {
                FeatureBitState::Defined
            }
        }
        FeatureBitState::Started => {
            if prev_window_signal_count >= SIGNAL_THRESHOLD {
                FeatureBitState::LockedIn
            } else if boundary_height >= d.timeout_height {
                FeatureBitState::Failed
            } else {
                FeatureBitState::Started
            }
        }
        FeatureBitState::LockedIn => FeatureBitState::Active,
        FeatureBitState::Active => FeatureBitState::Active,
        FeatureBitState::Failed => FeatureBitState::Failed,
    }
}

pub fn featurebit_state_at_height_from_window_counts(
    d: &FeatureBitDeployment,
    height: u64,
    window_signal_counts: &[u32],
) -> Result<FeatureBitEval, String> {
    if d.name.is_empty() {
        return Err("featurebits: name required".to_string());
    }
    if d.bit > 31 {
        return Err(format!("featurebits: bit out of range: {}", d.bit));
    }
    if d.timeout_height < d.start_height {
        return Err("featurebits: timeout_height < start_height".to_string());
    }

    let boundary_height = height - (height % SIGNAL_WINDOW);
    let target_boundary_index = boundary_height / SIGNAL_WINDOW;

    let need_windows = target_boundary_index as usize;
    if window_signal_counts.len() < need_windows {
        return Err(format!(
            "featurebits: need {} window_signal_counts entries, got {}",
            need_windows,
            window_signal_counts.len()
        ));
    }

    let mut state = FeatureBitState::Defined;
    for boundary_index in 0..=target_boundary_index {
        let bh = boundary_index * SIGNAL_WINDOW;
        let prev_cnt = if bh < SIGNAL_WINDOW {
            0
        } else {
            window_signal_counts[(boundary_index - 1) as usize]
        };
        state = next_state(state, bh, prev_cnt, d);
    }

    let prev_cnt = if boundary_height < SIGNAL_WINDOW {
        0
    } else {
        window_signal_counts[(target_boundary_index - 1) as usize]
    };

    Ok(FeatureBitEval {
        state,
        boundary_height,
        prev_window_signal_count: prev_cnt,
        signal_window: SIGNAL_WINDOW,
        signal_threshold: SIGNAL_THRESHOLD,
    })
}

File: clients/go/consensus/featurebits.go
package consensus

import "fmt"

type FeatureBitState string

const (
	FEATUREBIT_DEFINED   FeatureBitState = "DEFINED"
	FEATUREBIT_STARTED   FeatureBitState = "STARTED"
	FEATUREBIT_LOCKED_IN FeatureBitState = "LOCKED_IN"
	FEATUREBIT_ACTIVE    FeatureBitState = "ACTIVE"
	FEATUREBIT_FAILED    FeatureBitState = "FAILED"
)

type FeatureBitDeployment struct {
	Name          string
	Bit           uint8
	StartHeight   uint64
	TimeoutHeight uint64
}

type FeatureBitEval struct {
	State               FeatureBitState
	BoundaryHeight      uint64
	PrevWindowSignalCnt uint32
	SignalWindow        uint64
	SignalThreshold     uint32
}

func (d FeatureBitDeployment) Validate() error {
	if d.Bit > 31 {
		return fmt.Errorf("featurebits: bit out of range: %d", d.Bit)
	}
	if d.Name == "" {
		return fmt.Errorf("featurebits: name required")
	}
	if d.TimeoutHeight < d.StartHeight {
		return fmt.Errorf("featurebits: timeout_height < start_height")
	}
	return nil
}

func evalFeatureBitsNextState(
	prev FeatureBitState,
	boundaryHeight uint64,
	prevWindowSignalCount uint32,
	d FeatureBitDeployment,
) FeatureBitState {
	switch prev {
	case FEATUREBIT_DEFINED:
		if boundaryHeight >= d.StartHeight {
			return FEATUREBIT_STARTED
		}
		return FEATUREBIT_DEFINED
	case FEATUREBIT_STARTED:
		if prevWindowSignalCount >= SIGNAL_THRESHOLD {
			return FEATUREBIT_LOCKED_IN
		}
		if boundaryHeight >= d.TimeoutHeight {
			return FEATUREBIT_FAILED
		}
		return FEATUREBIT_STARTED
	case FEATUREBIT_LOCKED_IN:
		return FEATUREBIT_ACTIVE
	case FEATUREBIT_ACTIVE:
		return FEATUREBIT_ACTIVE
	case FEATUREBIT_FAILED:
		return FEATUREBIT_FAILED
	default:
		return prev
	}
}

func FeatureBitStateAtHeightFromWindowCounts(
	d FeatureBitDeployment,
	height uint64,
	windowSignalCounts []uint32,
) (FeatureBitEval, error) {
	if err := d.Validate(); err != nil {
		return FeatureBitEval{}, err
	}

	boundaryHeight := height - (height % SIGNAL_WINDOW)
	targetBoundaryIndex := boundaryHeight / SIGNAL_WINDOW

	needWindows := int(targetBoundaryIndex)
	if len(windowSignalCounts) < needWindows {
		return FeatureBitEval{}, fmt.Errorf(
			"featurebits: need %d window_signal_counts entries, got %d",
			needWindows,
			len(windowSignalCounts),
		)
	}

	state := FEATUREBIT_DEFINED
	for boundaryIndex := uint64(0); boundaryIndex <= targetBoundaryIndex; boundaryIndex++ {
		bh := boundaryIndex * SIGNAL_WINDOW
		var prevCnt uint32
		if bh < SIGNAL_WINDOW {
			prevCnt = 0
		} else {
			prevCnt = windowSignalCounts[boundaryIndex-1]
		}
		state = evalFeatureBitsNextState(state, bh, prevCnt, d)
	}

	var prevCnt uint32
	if boundaryHeight < SIGNAL_WINDOW {
		prevCnt = 0
	} else {
		prevCnt = windowSignalCounts[targetBoundaryIndex-1]
	}

	return FeatureBitEval{
		State:               state,
		BoundaryHeight:      boundaryHeight,
		PrevWindowSignalCnt: prevCnt,
		SignalWindow:        SIGNAL_WINDOW,
		SignalThreshold:     SIGNAL_THRESHOLD,
	}, nil
}
